src/System.Management.Automation/engine/CoreAdapter.cs

```diff
@@ -1693,13 +1696,19 @@ internal static Type EffectiveArgumentType(object arg)
                         return firstType.MakeArrayType();
                     }
                 }
-
-                return arg.GetType();
             }
-            else
+
+            return GetObjectType(arg, skipBasing: true);
+        }
+
+        internal static Type GetObjectType(object obj, bool skipBasing)
+        {
+            if (!skipBasing)
             {
-                return typeof(LanguagePrimitives.Null);
+                obj = PSObject.Base(obj);
             }
+
+            return obj == NullString.Value ? typeof(string) : obj?.GetType();
```

Comment:
I could imagine this logic being required in the future for different types, so hardcoding it here in the ternary does not look maintainable to me. How about making `NullString` implement something like `interface ITypeOverloadOverride { Type TypeForOverloadResolution`? Then you could replace this ternary with a generic `obj is ITypeOverloadOverride overridden ? overridden?.TypeForOverloadResolution : obj?.GetType();`

Reply:
> I could imagine this logic being required in the future for different types, so hardcoding it here in the ternary does not look maintainable to me. How about making `NullString` implement something like `interface ITypeOverloadOverride { Type TypeForOverloadResolution`? Then you could replace this ternary with a generic `obj is ITypeOverloadOverride overridden ? overridden?.TypeForOverloadResolution : obj?.GetType();`

`NullString` is a bit of a special case because it's the only type PowerShell tries to pretend as if it's a valuetype. I don't think it's likely we'll have to add another. Maybe would be a good issue to open but out of scope of this PR imo